src/main/java/com/spotify/confidence/EventSenderEngineImpl.java

```diff
@@ -1,162 +1,131 @@
-      while (true) {
-        try {
-          final String signal = uploadQueue.take();
-          final List<EventBatch> batches = List.copyOf(eventStorage.getBatches());
-          for (EventBatch batch : batches) {
-            final boolean uploadSuccessful = eventUploader.upload(batch).get();
-            if (uploadSuccessful) {
-              eventStorage.deleteBatch(batch.id());
-            }
-          }
-          if (signal.equals(SHUTDOWN_UPLOAD)) {
-            shutdownQueue.add(SHUTDOWN_UPLOAD_COMPLETED);
-          }
-        } catch (InterruptedException | ExecutionException e) {
-          throw new RuntimeException(e);
-        }
+      final var event = sendQueue.poll();
+      if (event != null) {
+        events.add(event);
+      } else {
+        if (intakeClosed) break;
+        LockSupport.parkUntil(Instant.now().plus(maxFlushInterval).toEpochMilli());
+      }
+      final boolean passedMaxFlushInterval =
+          !maxFlushInterval.isZero()
+              && Duration.between(latestFlushTime, Instant.now()).compareTo(maxFlushInterval) > 0;
+      if (events.size() == maxBatchSize || passedMaxFlushInterval) {
+        upload(events);
+        events = new ArrayList<>();
+        latestFlushTime = Instant.now();
       }
     }
+    upload(events);
   }
 
-  @Override
-  public void send(
-      String name, ConfidenceValue.Struct context, Optional<ConfidenceValue.Struct> message) {
-    if (!isStopped) {
-      log.trace("Sending event: {}", name);
-      writeQueue.add(
-          new Event(
-              EVENT_NAME_PREFIX + name,
-              message.orElse(ConfidenceValue.Struct.EMPTY),
-              context,
-              clock.currentTimeSeconds()));
-    }
-  }
-
-  @Override
-  public void close() {
-    // stop accepting new events
-    isStopped = true;
-    final ExecutorService thread = Executors.newSingleThreadExecutor();
-    thread.submit(
-        () -> {
-          try {
-            // wait until all the events in the queue are written
-            final String writeQueueShutdown = shutdownQueue.take();
-            assert (writeQueueShutdown.equals(SHUTDOWN_WRITE_COMPLETED));
-            // create the final batch
-            eventStorage.createBatch();
-            uploadQueue.add(SHUTDOWN_UPLOAD);
-            // wait until all the written events are uploaded
-            final String uploadQueueShutdown = shutdownQueue.take();
-            assert (uploadQueueShutdown.equals(SHUTDOWN_UPLOAD_COMPLETED));
-          } catch (InterruptedException e) {
-            throw new RuntimeException(e);
-          } finally {
-            try {
-              closeNow();
-            } catch (IOException e) {
-              // ignore
-            }
-          }
+  private void upload(List<Event> events) {
+    if (events.isEmpty()) return;
+    final CompletableFuture<Boolean> batchUploaded =
+        uploadExecutor.getStageAsync(() -> eventUploader.upload(new EventBatch(events)));
```

Comment:
What's the purpose of this separate executor? It's just doing non blocking io anyhow, why move it to another thread?

Reply:
`uploadExecutor` is the FailSafe executor so it does retries, and needs the ability to schedule retries. Right now it does this on the common pool.

Reply:
It's an executor from the `Failsafe` lib which handles retries with backoff internally.

Reply:
Makes sense, forgot about it being failsafe here